ocaml/parsing/parser.mly

```diff
@@ -3027,22 +3027,26 @@ local_strict_binding:
 ;
 match_case:
     pattern MINUSGREATER seq_expr
-      { Exp.case $1 $3 }
+      { Exp.case $1 (Case_rhs.simple $3) }
   | pattern WHEN seq_expr MINUSGREATER seq_expr
-      { Exp.case $1 ~guard:(Guard_predicate $3) $5 }
+      { Exp.case $1 (Case_rhs.boolean_guarded ~guard:$3 $5) }
   /* CR-soon rgodse: We should consider whether to also allow seq_expr, as this
      also parses without conflict.
 
      nroberts prefers expr, so that e1; e2 match p parses as e1; (e2 match p),
      if and when `e match p` is introduced as an expression
   */
-  | pattern WHEN expr MATCH pattern MINUSGREATER seq_expr
-      { let guard_pattern =
-          Guard_pattern.mk ~loc:(make_loc ($startpos($2), $endpos($5))) $3 $5
-        in
-        Exp.case $1 ~guard:(Guard_pattern guard_pattern) $7 }
+| pattern WHEN expr MATCH ioption(BAR) match_case
+      { Exp.case $1
+          (Case_rhs.pattern_guarded ~loc:(make_loc ($startpos($2), $endpos))
+             $3 [ $6 ]) }
+  | pattern WHEN expr MATCH LPAREN match_cases RPAREN
```

Comment:
Will this not be necessary if https://github.com/ocaml/ocaml/pull/12387 is merged? If so, I think this is fine, but we should maybe just leave a comment.

Reply:
I think it will. We require parens (as opposed to the optional ones around `match_cases` under the linked PR) to prevent the parser from parsing outer cases as cases of the pattern guard.

So for example `match e1 with p1 when e2 match p3 -> e3 | p4 -> e4` has two interpretations: 

(1)
```
match e1 with
| p1 when e2 match (p3 -> e3)
| p4 -> e4
```

(2)
```
match e1 with
| p1 when e2 match
    ( p3 -> e3
    | p4 -> e4
    )
```

The first highlighted case in `parser.mly` allows the example to parse successfully into (1), while the second highlighted case gives us a syntax for (2).

Reply:
Ok, it makes sense that both cases would be needed here to prevent the ambiguity being resolved in the wrong way. I suppose the only change after #12387 will be that you can add even more parens if you want.